### PR TITLE
assert that input and output buffers passed to transcode! are independent

### PR DESCRIPTION
**Pull request by baumgold**
As suggested by @quinnj in https://github.com/JuliaIO/TranscodingStreams.jl/pull/136#issuecomment-1504021093

### CODE REVIEW

**Comment by mkitti** [comment]
This is a good basic check, but I think we might want to actually check for aliasing. The `Base.dataids` seems to be a way to to do this. Maybe someone like @mbauman could help update us on the latest with regard to alias detection. My understanding from https://github.com/JuliaLang/julia/pull/25890 is that we could just see if the results of `Base.dataids` on each array are similar.

```julia
help?> Base.dataids
  Base.dataids(A::AbstractArray)

  Return a tuple of UInts that represent the mutable data segments of an array.

  Custom arrays that would like to opt-in to aliasing detection of their component parts can specialize this method to
  return the concatenation of the dataids of their component parts. A typical definition for an array that wraps a
  parent is Base.dataids(C::CustomArray) = dataids(C.parent).
```

For normal arrays, it seems 

```
julia> A = rand(1024);

julia> B = @view A[1:16];

julia> Base.dataids(A)
(0x0000014201b1a080,)

julia> Base.dataids(B)
(0x0000014201b1a080,)

julia> Base.dataids(A) == Base.dataids(B)
true

julia> pointer(A)
Ptr{Float64} @0x0000014201b1a080
```

File: src/transcode.jl [hunk 1]
```diff
@@ -117,6 +117,7 @@ function transcode!(
     codec::Codec,
     input::Buffer,
 )
+    @assert input.data !== output.data "input and outbut buffers must be independent"
```
Comment (mkitti) [inline src/transcode.jl:120]:
```suggestion
    @assert Base.dataids(input.data) !== Base.dataids(output.data) "input and outbut buffers must be independent"
```

**Review by quinnj — APPROVED** [review]
Thanks! Can we add a quick test too?

**Comment by mkitti** [comment]
@timholy just referred me to `Base.mightalias`.

```
help?> Base.mightalias
  Base.mightalias(A::AbstractArray, B::AbstractArray)

  Perform a conservative test to check if arrays A and B might share the same
  memory.

  By default, this simply checks if either of the arrays reference the same
  memory regions, as identified by their Base.dataids.
```

**Comment by mkitti** [comment]
I wonder if we are going to need an `unsafe_transcode!`. The alias detection might be too conservative.

File: src/noop.jl [hunk 2]
```diff
@@ -116,16 +116,15 @@ function Base.unsafe_write(stream::NoopStream, input::Ptr{UInt8}, nbytes::UInt)
     end
 end
 
-function Base.transcode(::Type{Noop}, data::ByteData)
-    return transcode(Noop(), data)
-end
+initial_output_size(codec::Noop, input::Memory) = length(input)
```
Comment (mkitti) [inline src/noop.jl:119]: Do we still need this?

**Comment by baumgold** [comment]
> I wonder if we are going to need an `unsafe_transcode!`. The alias detection might be too conservative.

Agreed, added.

**Review by quinnj — APPROVED** [review]

**Pull request merged**